public/js/ace-1.2.5/worker-latex.js

```diff
@@ -1414,67 +1414,412 @@ var Mirror = require("../worker/mirror").Mirror;
+	var end_col = toEnd - linePosition[toLine] + 1;
+
+	errors.push({row: line,
+		     column: start_col,
+		     start_row: fromLine,
+		     start_col: start_col,
+		     end_row: toLine,
+		     end_col: end_col,
+		     type:"error",
+		     text:message,
+		     suppressIfEditing:true});
+    };
+
+
+    var EnvErrorFromTo = function (fromEnv, toEnv, message, options) {
+	if(!options) { options = {} ; };
+	var fromToken = fromEnv.token, toToken = toEnv.closeToken || toEnv.token;
+	var fromLine = fromToken[0], fromStart = fromToken[2], fromEnd = fromToken[3], fromSeq = fromToken[4];
+	if (!toToken) {toToken = fromToken;};
+	var toLine = toToken[0], toStart = toToken[2], toEnd = toToken[3], toSeq = toToken[4];
+	if (!toEnd) { toEnd = toStart + 1;};
+	var start_col = fromStart - linePosition[fromLine];
+	var end_col = toEnd - linePosition[toLine] + 1;
+	errors.push({row:toLine,
+		     column:end_col,
+		     start_row:fromLine,
+		     start_col: start_col,
+		     end_row:toLine,
+		     end_col: end_col,
+		     type:"error",
+		     text:message,
+		     suppressIfEditing:options.suppressIfEditing});
+    };
+
+    var EnvErrorTo = function (toEnv, message) {
+	var token = toEnv.closeToken || toEnv.token;
+	var line = token[0], type = token[1], start = token[2], end = token[3], seq = token[4];
+	if (!end) { end = start + 1; };
+	var end_col = end - linePosition[line] + 1;
+	var err = {row: line,
+		   column: end_col,
+		   start_row:0,
+		   start_col: 0,
+		   end_row: line,
+		   end_col: end_col,
+		   type:"error",
+		   text:message};
+	errors.push(err);
+    };
+
+    var EnvErrorFrom = function (env, message) {
+	var token = env.token;
+	var line = token[0], type = token[1], start = token[2], end = token[3], seq = token[4];
+	var start_col = start - linePosition[line];
+	var end_col = Infinity;
+	errors.push({row: line,
+		     column: start_col,
+		     start_row:line,
+		     start_col: start_col,
+		     end_row: lineNumber,
+		     end_col: end_col,
+		     type:"error",
+		     text:message});
+    };
+
+    var checkingDisabled = false;
+
+    while (true) {
+	var result = SPECIAL.exec(text);
+	if (result == null) {
+	    if (idx < text.length) {
+		Tokens.push([lineNumber, "Text", idx, text.length]);
+	    }
+	    break;
+	}
+	if (result && result.index <= pos) {
+	    break;
+	};
+	pos = result.index;
+	var newIdx = SPECIAL.lastIndex;
+	if (pos > idx) {
+	    Tokens.push([lineNumber, "Text", idx, pos]);
+	}
+	for (var i = idx; i < pos; i++) {
+	    if (text[i] === "\n") {
+		lineNumber++;
+		linePosition[lineNumber] = i+1;
+	    }
+	}
+	idx = newIdx;
+	var code = result[0];
+	if (code === "%") {
+	    var newLinePos = text.indexOf("\n", idx);
+	    if (newLinePos === -1) {
+		newLinePos = text.length;
+	    };
+	    var commentString = text.substring(idx, newLinePos);
+	    if (commentString.indexOf("%novalidate") === 0) {
+		return [];
+	    } else if(!checkingDisabled && commentString.indexOf("%begin novalidate") === 0) {
```

Comment:
this big if else statment would be more redable if ti was
```

if(isComment){
    processComment()
}else if(isNoValidate){
    processNoValidate()
}...etc
```

Just to break it down into more bit size chucks for other peoples head